src/jit/morph.cpp

```diff
@@ -11457,7 +11457,20 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
             break;
 
         case GT_DIV:
-
+            // Replace "val / dcon" with "val * (1.0 / dcon)" if dcon is a power of two.
+            // Powers of two within range are always exactly represented,
+            // so multiplication by the reciprocal is safe in this scenario
+            if (op2->IsCnsFltOrDbl())
+            {
+                double divisor = op2->AsDblCon()->gtDconVal;
+                if (((typ == TYP_DOUBLE) && FloatingPointUtils::hasPreciseReciprocal(divisor)) ||
+                    ((typ == TYP_FLOAT) && FloatingPointUtils::hasPreciseReciprocal(forceCastToFloat(divisor))))
+                {
+                    oper = GT_MUL;
+                    tree->ChangeOper(oper);
+                    op2->AsDblCon()->gtDconVal = 1.0 / divisor;
```

Comment:
It might be worth noting that this is safe and doing the single operation in single precision isn't required.

The paper `Innocuous Double Rounding of Basic Arithmetic Operations` provides a proof that a single primitive operation done to at least twice the precision of the target format does not incur error due to double-rounding (hence for `float divisor` doing `(float)(1.0 / divisor)` is the same as `(1.0f / divisor)`).

This is not a safe thing to do across multiple operations (you must downcast back to float after each individual operation) nor is it safe if one of the inputs could not be exactly represented as a `float` (e.g. if you have `double divisor` and `(float)divisor != divisor`).